d2layouts/d2sequence/layout.go

```diff
@@ -4,90 +4,135 @@ import (
 	"context"
 	"fmt"
 	"math"
+	"sort"
 
 	"oss.terrastruct.com/d2/d2graph"
 	"oss.terrastruct.com/d2/lib/geo"
 	"oss.terrastruct.com/d2/lib/go2"
 	"oss.terrastruct.com/d2/lib/label"
+	"oss.terrastruct.com/d2/lib/shape"
 )
 
 func Layout(ctx context.Context, g *d2graph.Graph) (err error) {
-	edgeYStep := MIN_EDGE_DISTANCE
-	actorXStep := MIN_ACTOR_DISTANCE
-	maxActorHeight := 0.
+	sd := &sequenceDiagram{
+		graph:          g,
+		objectRank:     make(map[*d2graph.Object]int),
+		objectDepth:    make(map[*d2graph.Object]int),
+		minEdgeRank:    make(map[*d2graph.Object]int),
+		maxEdgeRank:    make(map[*d2graph.Object]int),
+		edgeYStep:      MIN_EDGE_DISTANCE,
+		actorXStep:     MIN_ACTOR_DISTANCE,
+		maxActorHeight: 0.,
+	}
+
+	sd.init()
+	sd.placeActors()
+	sd.placeActivationBoxes()
+	sd.routeEdges()
+	sd.addLifelineEdges()
+
+	return nil
+}
+
+type sequenceDiagram struct {
+	graph *d2graph.Graph
+
+	edges       []*d2graph.Edge
+	actors      []*d2graph.Object
+	activations []*d2graph.Object
+
+	// can be either actors or activation boxes
+	objectRank  map[*d2graph.Object]int
+	objectDepth map[*d2graph.Object]int
+
+	// keep track of the first and last edge of a given actor
+	// needed for activation boxes
+	minEdgeRank map[*d2graph.Object]int
+	maxEdgeRank map[*d2graph.Object]int
 
-	actorRank := make(map[*d2graph.Object]int)
-	for rank, actor := range g.Objects {
-		actorRank[actor] = rank
+	edgeYStep      float64
+	actorXStep     float64
+	maxActorHeight float64
+}
+
+func (sd *sequenceDiagram) init() {
+	sd.edges = make([]*d2graph.Edge, len(sd.graph.Edges))
+	copy(sd.edges, sd.graph.Edges)
+
+	queue := make([]*d2graph.Object, len(sd.graph.Root.ChildrenArray))
+	copy(queue, sd.graph.Root.ChildrenArray)
+	for len(queue) > 0 {
+		obj := queue[0]
+		queue = queue[1:]
+
+		if sd.isActor(obj) {
+			sd.actors = append(sd.actors, obj)
+			sd.objectRank[obj] = len(sd.actors)
+			sd.objectDepth[obj] = 0
+			sd.maxActorHeight = math.Max(sd.maxActorHeight, obj.Height+VERTICAL_PAD)
+		} else {
+			// activations boxes are always rectangles and have no labels
+			obj.Attributes.Label = d2graph.Scalar{Value: ""}
+			obj.Attributes.Shape = d2graph.Scalar{Value: shape.SQUARE_TYPE}
```

Comment:
we should enforce this in the compiler, help users out with error messages. https://github.com/terrastruct/d2/issues/268

Reply:
ok, will get it..